picard/item.py

```diff
@@ -180,3 +183,110 @@ def keep_original_images(self):
                 file.keep_original_images()
         self.enable_update_metadata_images(True)
         self.update_metadata_images()
+
+    def children_metadata_items(self):
+        """Yield MetadataItems that are children of the current object"""
+
+    def iter_children_items_metadata(self, metadata_attr):
+        for s in self.children_metadata_items():
+            if metadata_attr in s.iter_children_items_metadata_ignore_attrs:
+                continue
+            yield getattr(s, metadata_attr)
+
+    @staticmethod
+    def get_sources_metadata_images(sources_metadata):
+        images = set()
+        for s in sources_metadata:
+            images = images.union(s.images)
+        return images
+
+    def remove_metadata_images_from_children(self, removed_sources):
+        """Remove the images in the metadata of `removed_sources` from the metadata.
+
+        Args:
+            removed_sources: List of child objects (`Track` or `File`) which's metadata images should be removed from
+        """
+        changed = False
+
+        for metadata_attr in self.update_children_metadata_attrs:
+            removed_images = self.get_sources_metadata_images(getattr(s, metadata_attr) for s in removed_sources)
+            sources_metadata = list(self.iter_children_items_metadata(metadata_attr))
+            metadata = getattr(self, metadata_attr)
+            changed |= metadata.remove_images(sources_metadata, removed_images)
+
+        return changed
+
+    def add_metadata_images_from_children(self, added_sources):
+        """Add the images in the metadata of `added_sources` to the metadata.
+
+        Args:
+            added_sources: List of child objects (`Track` or `File`) which's metadata images should be added to current object
+        """
+        changed = False
+
+        for metadata_attr in self.update_children_metadata_attrs:
+            added_images = self.get_sources_metadata_images(getattr(s, metadata_attr) for s in added_sources)
+            metadata = getattr(self, metadata_attr)
+            changed |= metadata.add_images(added_images)
+
+        return changed
+
+    def update_metadata_images_from_children(self):
+        """Update the metadata images of the current object based on its children.
+
+        Based on the type of the current object, this will update `self.metadata.images` to
+        represent the metadata images of all children (`Track` or `File` objects).
+
+        This method will iterate over all children and completely rebuild
+        `self.metadata.images`. Whenever possible the more specific functions
+        `add_metadata_images_from_children` or `remove_metadata_images_from_children` should be used.
+
+        Returns:
+            bool: True, if images where changed, False otherwise
+        """
+        from picard.util.imagelist import ImageList
+
+        class ImageListState:
```

Comment:
I'd move the class declaration outside the function. I think we should avoid class and function declaration inside functions except for very simple cases. But this is complex enough to stay on it's own. If we really want to scope it further then just the module we could put it inside the Item class.

Reply:
I'll move it once this is merged.